File: plugins/lofi/tape-stop/tape-stop.js
// Tape Stop — variable-speed playback from a continuously-rolling
// buffer. Write head always advances at 1× (capturing live audio); the
// read head advances at the user-set speed. At speed=1 it sounds like a
// short delay, at speed→0 the playback grinds to a halt with the
// pitch dropping accordingly — exactly the tape-deck-pulling-the-plug
// effect heard on countless EDM drops.

function TapeStopProcessor() {
  // 2 s @ 48 kHz is plenty of buffer for slowdown trails.
  this.buf = new Array(96000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.bufLen = this.buf.length;
  this.writeIdx = 0;
  this.readPos = 0;
}
TapeStopProcessor.parameterDescriptors = [
  { name: 'speed', label: 'Speed', defaultValue: 1, minValue: 0, maxValue: 1 },
  { name: 'mix',   label: 'Mix',   defaultValue: 1, minValue: 0, maxValue: 1 }
];
TapeStopProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var speed = parameters.speed[0];
  var mix = parameters.mix[0];
  for (var i = 0; i < input.length; i++) {
    this.buf[this.writeIdx] = input[i];
    this.writeIdx = (this.writeIdx + 1) % this.bufLen;
    var idx = Math.floor(this.readPos) % this.bufLen;
    if (idx < 0) idx += this.bufLen;
    var i1 = (idx + 1) % this.bufLen;
    var frac = this.readPos - Math.floor(this.readPos);
    var wet = this.buf[idx] * (1 - frac) + this.buf[i1] * frac;
    output[i] = input[i] * (1 - mix) + wet * mix;
    this.readPos += speed;
    if (this.readPos >= this.bufLen) this.readPos -= this.bufLen;
  }
  return true;
};

registerProcessor('tape-stop', TapeStopProcessor);

File: plugins/lofi/tape-stop/TapeStop.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Tape Stop (native port) — variable-speed playback from a rolling
// buffer. speed → 0 grinds the audio to a halt with the pitch dropping
// with it. The classic EDM-drop tape-pull effect.
public final class TapeStop implements VocalMonitorNativePlugin {
    private float[] buf;
    private int bufLen;
    private int writeIdx = 0;
    private float readPos = 0f;
    private float speed = 1f, mix = 1f;

    @Override
    public void init(int sr) {
        bufLen = 96000;
        buf = new float[bufLen];
        writeIdx = 0;
        readPos = 0f;
    }

    @Override public String[] parameterNames() { return new String[] { "speed", "mix" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) { return 1f; }
    @Override public String parameterLabel(String n) { return "speed".equals(n) ? "Speed" : "Mix"; }
    @Override public void setParameter(String n, float v) {
        if ("speed".equals(n)) speed = v;
        else if ("mix".equals(n)) mix = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float spd = speed;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float[] b = buf;
        final int bL = bufLen;
        int w = writeIdx;
        float r = readPos;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            b[w] = input[i];
            w++; if (w >= bL) w = 0;
            int idx = (int) Math.floor(r) % bL;
            if (idx < 0) idx += bL;
            int i1 = idx + 1; if (i1 >= bL) i1 = 0;
            float frac = r - (float) Math.floor(r);
            float wet = b[idx] * (1f - frac) + b[i1] * frac;
            output[i] = input[i] * dry + wet * mixLocal;
            r += spd;
            if (r >= bL) r -= bL;
        }
        writeIdx = w;
        readPos = r;
    }
}
